src/radical/pilot/agent/agent_0.py

```diff
@@ -532,13 +535,134 @@ def _check_commands(self):
         return True
 
 
+    # --------------------------------------------------------------------------
+    #
+    def _check_rpc(self):
+        '''
+        check if the DB has any RPC request for this pilot.  If so, then forward
+        that request as `rpc_req` command on the CONTROL channel, and listen for
+        an `rpc_res` command on the same channel, for the same rpc id.  Once
+        that response is received (from whatever component handled that
+        command), send the response back to the databse for the callee to pick
+        up.
+        '''
+
+        # FIXME: implement a timeout, and/or a registry of rpc clients
+
+        self._log.debug('=== rpc check')
+
+        retdoc = self._dbs._c.find_and_modify(
+                    query ={'uid' : self._pid},
+                    fields=['rpc_req'],
+                    update={'$set': {'rpc_req': None}})
+
+        if not retdoc:
+            # no rpc request found
+            return True
+
+        rpc_req = retdoc.get('rpc_req')
+        if rpc_req is None:
+            # document has no rpc request
+            return True
+
+        self._log.debug('=== rpc req: %s', rpc_req)
+
+        # RPCs are synchronous right now - we send the RPC on the command
+        # channel, hope that some component picks it up and replies, and then
+        # return that reply.  The reply is received via a temporary callback
+        # defined here, which will receive all CONTROL messages until the right
+        # rpc response comes along.
+        def rpc_cb(topic, msg):
+
+            rpc_id  = rpc_req['uid']
+
+            cmd     = msg['cmd']
+            rpc_res = msg['arg']
+
+            if cmd != 'rpc_res':
+                # not an rpc responese
+                return True
+
+            if rpc_res['uid'] != rpc_id:
+                # not the right rpc response
+                return True
+
+            # send the response to the DB
+            self._dbs._c.update({'type'  : 'pilot',
+                                 'uid'   : self._pid},
+                                {'$set'  : {'rpc_res': rpc_res}})
+
+            # work is done - unregister this temporary cb (rpc_cb)
+            return False
+
+
+        self.register_subscriber(rpc.CONTROL_PUBSUB, rpc_cb)
+
+        # ready to receive and proxy rpc response -- forward rpc request on
+        # control channel
+        self.publish(rpc.CONTROL_PUBSUB, {'cmd' : 'rpc_req',
+                                          'arg' : rpc_req})
+
+        return True  # keeb cb registered (self._check_rpc)
+
+
+    # --------------------------------------------------------------------------
+    #
+    def _check_control(self, _, msg):
```

Comment:
Would you mind if I try to create a unit test for this method? I think I have an idea of how to do it.

Reply:
> Would you mind if I try to create a unit test for this method?

Would I mind?   WOULD *I* MIND?  I thought you'd knew me better than that... :-P

Reply:
Done